File: 1519-number-of-nodes-in-the-sub-tree-with-the-same-label/1519-number-of-nodes-in-the-sub-tree-with-the-same-label.java
class Solution {
    Map<Integer,int[]> cache = new HashMap();
    public int[] countSubTrees(int n, int[][] edges, String labels) {
        int[] rc = new int[n];
        
        Map<Integer,List<Integer>> map = new HashMap();
        for (int i=0;i<n;i++)
            map.put(i,new ArrayList());
        for (int[] e : edges) {
            map.get(e[0]).add(e[1]);
            map.get(e[1]).add(e[0]);
        }
            
        for (int i=0;i<n;i++)
        {
            int[] temp = dfs(i,map,labels, new HashSet());
            rc[i] = temp[labels.charAt(i)-'a'];
        }
            
        return rc;
    }
    int[] dfs (int i, Map<Integer,List<Integer>> map, String labels, Set<Integer> set) {
        if ( set.contains(i) ) return new int[26];
        if ( cache.containsKey(i) ) return cache.get(i);
        int[] temp = new int[26];
        temp[ labels.charAt(i) - 'a']++;
        set.add(i);
        
        for (int neigh : map.get(i) ) {
            int[] t = dfs(neigh,map,labels,set);
            for (int j=0;j<26;j++)
                temp[j] += t[j];
        }
        
        cache.put(i,temp);
        return temp;
        
    }
}

File: 1519-number-of-nodes-in-the-sub-tree-with-the-same-label/1519-number-of-nodes-in-the-sub-tree-with-the-same-label.cpp
class Solution {
public:
    unordered_map<int, vector<int>> graph;
    unordered_map<int, char> conn;
    unordered_set<int> visited;
    void createGraph(vector<vector<int>>& edges){
        for(auto &v : edges){
            graph[v[0]].push_back(v[1]);
            graph[v[1]].push_back(v[0]);
        }
    }
    vector<int> dfs(int node, unordered_set<int> &visited, vector<int> & ans){
        if(visited.count(node)){
            vector<int> re;
            return re;
        }
        visited.insert(node);
        vector<int> c(26, 0);
        for(auto &nei : graph[node]){
            vector<int> temp = dfs(nei, visited, ans);
            if (temp.size()){
                for(int i = 0; i < 26; ++i){
                    c[i] += temp[i];
                }

            }
        }
        c[conn[node]-'a']++;
        ans[node] = c[conn[node]-'a'];
        return c;
    }
    vector<int> countSubTrees(int n, vector<vector<int>>& edges, string labels) {
        createGraph(edges);

        for(int i = 0; i < labels.size();++i){
            conn[i] = labels[i];
        }
        vector<int> ans(n);
        dfs(0, visited, ans);
        return ans;

    }
};